core/src/main/scala/kafka/cluster/Partition.scala

```diff
@@ -186,8 +186,12 @@ class Partition(val topic: String,
       // record the epoch of the controller that made the leadership decision. This is useful while updating the isr
       // to maintain the decision maker controller's epoch in the zookeeper path
       controllerEpoch = partitionStateInfo.controllerEpoch
-      // add replicas that are new
-      allReplicas.foreach(replica => getOrCreateReplica(replica))
+      // add replicas that are new and reset lastCatchUpTime of non-isr replicas
+      allReplicas.foreach(replica => {
+        val r = getOrCreateReplica(replica)
+        if (!partitionStateInfo.isr.contains(replica))
+          r.resetLastCatchUpTime()
+      })
```

Comment:
If the broker is already the leader of this partition, then the broker already have up-to-date lastCaughtUpTime of each follower. If the broker was not the leader of this partition, in the worst case the isr will temporarily drop below min isr of this partition, making this partition unavailable for produce operation. The producer will retry upon NotEnoughReplicasException and should succeed soon. It seems OK, because when there is leadership change from one broker to another, producer need to update metadata and retry anyway.

On the other hand, if we set lastCaughtUpTime to be now, then in the worst case a replica will stay in ISR for `2 x ReplicaLagTimeMaxMs`, which seems worse because it breaks the semantics of ReplicaLagTimeMaxMs (in addition to how we execute `maybeShrinkIsr()`).

What do you think? I can set lastCaughtUpTime to now if you think that is better.